services/web/server/src/simcore_service_webserver/projects/projects_utils.py

```diff
@@ -223,3 +226,58 @@ async def project_get_depending_nodes(
 def extract_dns_without_default_port(url: URL) -> str:
     port = "" if url.port == 80 else f":{url.port}"
     return f"{url.host}{port}"
+
+
+def any_node_inputs_changed(
+    updated_project: ProjectDict, current_project: ProjectDict
+) -> bool:
+    """Returns true if any change is detected in the node inputs of the updated project
+
+    Based on the limitation we are detecting with this check, new nodes only account for
+    a "change" if they add link inputs.
+    """
+    # NOTE: should not raise exceptions in production
+
+    project_uuid = current_project["uuid"]
+
+    assert (  # nosec
+        updated_project.get("uuid") == project_uuid
+    ), f"Expected same project, got {updated_project.get('uuid')}!={project_uuid}"
+
+    assert (  # nosec
+        "workbench" in updated_project
+    ), f"expected validated model but got {list(updated_project.keys())=}"
+
+    assert (  # nosec
+        "workbench" in current_project
+    ), f"expected validated model but got {list(current_project.keys())=}"
+
+    # detect input changes in existing nodes
+    for node_id, updated_node in updated_project["workbench"].items():
+        if current_node := current_project["workbench"].get(node_id, None):
+            if (updated_inputs := updated_node.get("inputs")) != current_node.get(
+                "inputs"
+            ):
+                log.debug(
+                    "Change detected in projects[%s].workbench[%s].%s",
+                    f"{project_uuid=}",
+                    f"{node_id=}",
+                    f"{updated_inputs=}",
+                )
+                return True
+
+        else:
+            # for new nodes, detect only added link
+            for input_name, input_value in updated_node.get("inputs", {}).items():
+                # TODO: how to ensure this list of "links types" is up-to-date!??
+                # These types represent links that node-ports need to handle
+                if isinstance(input_value, PortLink, DownloadLink, BaseFileLink):
```

Comment:
Maybe this should be provided by some utility function in nodeports and have some tests which makes sure the list is up to date. But it will still not be prefect.

Reply:
the tests are fine but I would rather have something closer to that line  ... I added some ``assertion``s next to t but the check is for the moment limited.